subprojects/core/src/main/java/org/gradle/api/file/DirectoryVar.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.provider.PropertyState;
+import org.gradle.api.provider.Provider;
+
+import java.io.File;
+
+/**
+ * Represents some configurable directory location, whose value is mutable and is not necessarily currently known until later.
+ * <p>
+ * <b>Note:</b> This interface is not intended for implementation by build script or plugin authors. An instance of this class can be created
+ * using the {@link ProjectLayout#newDirectoryVar()} method.
+ *
+ * @since 4.1
+ */
+@Incubating
+public interface DirectoryVar extends Provider<Directory>, PropertyState<Directory> {
```

Comment:
I'm a bit uneasy if the name for this interface (and `RegularFileVar`). I fail to see the reason of using suffix `Var` to the interface name. I feel like `ConfigurableDirectory` (and `ConfigurableRegularFile`) would be more appropriate. I base myself on existing name such as `ConfigurableFileCollection` and `ConfigurableFileTree`.